### ISSUE TITLE
Benchmark case studies

### ISSUE THREAD

**Issue by viktor61** [scikit-optimize/scikit-optimize#180]
Hi all,

I met Tim (betatim) today who told me about this very interesting project, and I would like to contribute.

I have worked a bit on optimisation before, and have benchmarked some libraries like hyperopt 
to compare their performance when used on real ML problems and complicated theoretical functions. Tim told me this is something this project would need.

I am happy to benchmark scikit-optimise on some nasty optimisation problems, and also apply it on a real dataset using popular techniques with lots of hyperparameters like XGBoost. And of course compare the performance vs other libraries. 

Let me know if this is something you would be interested.

Cheers,
Victor

**Comment by MechCoder** [comment]
That would be really really great. Thanks!

**Comment by glouppe** [comment]
+1, this would be really great!

As I was suggesting to Tim, we should however keep in mind that functions with very weird surfaces are not necessarily those that we mainly target. We rather aim at functions that are expensive to evaluate, but for which the surface can in fact be "well-behaved". 

In any case, it would not harm to some more in `skopt.benchmarks` and better benchmarks all variants of the algorithm.

**Comment by viktor61** [comment]
@glouppe I totally agree with you; the main criteria is to assess its performance on functions expensive to evaluate, and see how fast it converges toward the optimum.

I also usually apply the algorithm on a function with lots of local minima, yet fast to evaluate. The purpose is mainly to see how robust the algorithm is against lots of local minima. I usually use [Ackley function](http://www.sfu.ca/~ssurjano/ackley.html). If it does not pass this test it is likely to perform poorly on a real problem (in my opinion).

That being said, let's focus on a relevant case study first :) I am thinking about using a Kaggle dataset, for instance the one from the homesite competition that I entered. 

Would that be a suitable dataset for a benchmark?

**Comment by betatim** [comment]
I think it would be good to have the benchmark scripts/notebook as part of the repository or even part of the unit tests (similar to the `branin` etc tests we have right now). This might place some limits on what datasets we can use (they have to be public?).

As I see it right now there are at least two tasks:

* add more exotic yet relevant functions to `skopt.benchmarks`
* develop a notebook to benchmark real world problems/datasets

**Comment by MechCoder** [comment]
@viktor61 While you are it, would you mind testing for different values of the base estimator in `forest_minimize`? (https://github.com/scikit-optimize/scikit-optimize/issues/132) . It would be interesting to see if it has huge impacts.

**Comment by viktor61** [comment]
Just to let you know I was familiarising myself with the code. I should be able to commit some new benchmark functions next week, and @MechCoder I will have a look at you request as well.

**Comment by betatim** [comment]
I found [this set](https://github.com/sigopt/evalset) of test functions together with a publication http://arxiv.org/abs/1603.09441

SMAC doesn't perform very well in their benchmark which the authors say is because none of the problems are hard enough (not high dimensionality, categorical variables). We still need to find benchmarks for those.

**Comment by viktor61** [comment]
Thanks @betatim I hadn't seen that one.

Truth is all those artificial functions are only interesting for their short evaluation time.

If we really want to demonstrate the usefulness for practical problems we will need a real dataset. In my opinion, the best strategy to advertise the library would be to do a benchmark on one of the current Kaggle competition and post it on their forum :)

**Comment by betatim** [comment]
Another set of benchmarks (on real datasets): https://github.com/rhiever/sklearn-benchmarks. Needs some reading to understand how to use them.

Agreed that real datasets are the way to go. The drawback with them is that they take a long time to run and that you don't know what the true minimum is.

In the end having both is best (it is ensembles all the way down 😀)

**Comment by betatim** [comment]
Started a notebook benchmarking stuff: https://gist.github.com/betatim/d37c5737444156240e267e2d28593dce needs a modified `test_funcs.py` which is included in the gist.

**Comment by MechCoder** [comment]
Thanks for starting @betatim . Will have a look now.

**Comment by MechCoder** [comment]
We should cache the results in order to prevent the pain of rerunning when we try out new stuff.

**Comment by MechCoder** [comment]
See; https://github.com/scikit-optimize/scikit-optimize/issues/201
